File: internal/spec/edits.go
package spec

// derivedEdit is a quoter's in-place change to a quoted message, located in the
// rows it was found among: the COPY the quoter actually pasted (a DERIVED entry,
// decided at ingest — see twins.go), the BASE it modified, and the HOST it was
// sighted inside — the message that re-quoted and therefore carries the change.
//
// All three are the same three rows the relation already names; what this adds
// is only the resolution of the host, which the corpus records as a sighting
// rather than as a column.
type derivedEdit struct {
	Copy *entryRow
	Base *entryRow
	Host *entryRow
}

// derivedEdits finds every quoter's edited copy of a quoted message among rows.
//
// The relation is DECIDED at ingest (FindDerived classifies a modified re-quote
// as derived, not a twin and not unrelated); this only surfaces it, and it is
// shared so that the two renderers of one corpus surface it the same way. A page
// build draws the edit inside the host's bubble (see generate.go's attachEdits);
// the reading pane draws the same bubble over /v1/chains, and a second copy of
// these rules is how the two would come to disagree about whether a message
// carries an edit at all.
//
// Both the base and the host must be among rows. When either is not — the change
// is anchored outside this selection, or the quoting message was never collected
// — the derived copy keeps its own row and link and no edit is attached, so the
// trail is not silently lost. The host is the FIRST sighting that is present,
// which is what makes the choice stable across reads: rows are walked in the
// same order every time, and a copy sighted in several messages is shown in one
// of them rather than in all.
func derivedEdits(rows []*entryRow) []derivedEdit {
	at := map[int64]*entryRow{}
	for _, r := range rows {
		at[r.ID] = r
	}
	var out []derivedEdit
	for _, r := range rows {
		if !r.Derived || r.ParentID == 0 {
			continue // not an edited copy of anything
		}
		base, ok := at[r.ParentID]
		if !ok {
			continue // the base it changed is not here
		}
		var host *entryRow
		for _, h := range r.SeenIn {
			if x, ok := at[h]; ok && x != r {
				host = x
				break
			}
		}
		if host == nil {
			continue // the quoting message is not here either
		}
		out = append(out, derivedEdit{Copy: r, Base: base, Host: host})
	}
	return out
}

File: src/lib/edits.ts
/**
 * edits — turn a quoter's edited copy of a quoted message into what a bubble
 * draws (issue #42).
 *
 * The corpus decides at ingest that a re-quoted message with words inserted
 * inside it is a MODIFIED COPY of the message it quotes, and names the base it
 * modified. What a renderer does with that is the same in both places that draw
 * a message: the change is drawn inline inside the message that made it, diffed
 * against the original, attributed to the quoting message's own sender and clock
 * — never left floating as its own node.
 *
 * That rule is here, once, because the two callers hold different things. A page
 * build resolves ids against the spec's own rows; the reading pane resolves them
 * against the corpus entries the thread handed it. Which entries an edit names,
 * what the diff is made of, and what the attribution falls back to are the same
 * questions in both, so only the naming is the caller's: `find` says how to read
 * a person and a clock in its own vocabulary.
 */
import { editHtml } from "./editDiff";

/** A quoter's edit resolved for the bubble: diff markup plus attribution. */
export interface RowEdit {
  /** the id of the original message the change was made to (the anchor target),
   *  in whatever id space the caller's own bubbles carry */
  base: string;
  who: string;
  time: string;
  /** the quoter's modified text as diff-marked HTML (`.edel` strike / `.eins` insert) */
  html: string;
  /** the original message's sender, for "original from <y>" (may be empty) */
  origWho: string;
  /** the original message's date + time, for "at <timestamp>" (may be empty) */
  origStamp: string;
}

/** One edit as it arrives: the derived copy's id, the base it modified, and —
 *  where the writer stated them — who made the change, when, and the modified
 *  text. A page build fills all three in; a trail read leaves who and when to
 *  the bubble the edit is drawn inside. */
export interface EditRef {
  id?: string;
  base?: string;
  who?: string;
  time?: string;
  body?: string;
}

/** What a renderer must know about the entries an edit names, in its own
 *  vocabulary: the entry's rendered markup (a pasted quote's formatting lives in
 *  the copy, not in the bare original), and how this caller writes the entry's
 *  author and clock. */
export interface EditEntry {
  html: string;
  who: string;
  stamp: string;
}

/**
 * Resolve the edits attached to one message into what its bubble draws.
 *
 * `host` is the message the edit was made in — its own sender and clock are the
 * attribution, because the person who edited the quote is the person who sent
 * the message quoting it, whatever the recovered copy still says its sender was.
 * The diff is the quoter's stored text against the original's rendered body: one
 * plain text, one markup, which is what editHtml reads.
 */
export function resolveEdits(
  edits: EditRef[] | undefined,
  find: (id: string | undefined) => EditEntry | undefined,
  host: { who: string; time: string },
): RowEdit[] | undefined {
  if (!edits?.length) return undefined;
  return edits.map((ed) => {
    const copy = find(ed.id);
    const base = find(ed.base);
    return {
      base: ed.base ?? "",
      who: ed.who || host.who,
      time: ed.time || host.time,
      // The copy's markup is the surface: it is where the quoter's own formatting
      // survived (an answer written in red inside a forwarded thread), and the
      // words it does not hold are the ones the diff marks.
      html: editHtml(copy?.html ?? "", base?.html ?? "", ed.body ?? ""),
      origWho: base?.who ?? "",
      origStamp: base?.stamp ?? "",
    };
  });
}
